py/dml/ctree.py

```diff
@@ -1418,6 +1413,31 @@ def mkEquals(site, lh, rh):
     else:
         return Equals.make(site, lh, rh)
 
+class IdentityEq(Expression):
+    priority = dml.expr.Apply.priority
+    type = TBool()
+
+    @auto_init
+    def __init__(self, site, lh, rh): pass
+
+    def __str__(self):
+        lh = str(self.lh)
+        rh = str(self.rh)
+        if self.lh.priority <= Equals.priority:
+            lh = '('+lh+')'
+        if self.rh.priority <= Equals.priority:
+            rh = '('+rh+')'
+        return lh + ' == ' + rh
+
+    def read(self):
+        if self.site is None or isinstance(self.site, SimpleSite):
+            return f'_identity_eq({self.lh.read()}, {self.rh.read()})'
+        else:
+            return (f'_identity_eq_at_site({self.lh.read()}, {self.rh.read()}'
```

Comment:
I'd invert the logic, if `isinstance(self.site, DumpableSite)` then we can use the operations provided by DumpableSite.

Reply:
I _was_ about to say "but it's `SimpleSite` that's the outlier in not supporting `filename()` and `lineno`", but now actually I can see that it does have simplistic definitions for those. Because of that I'd rather get rid of the `isinstance` check entirely and just keep the `is None`.